backend/controllers/searchOffenders.js

```diff
@@ -13,65 +13,88 @@ router.get('/', asyncMiddleware(async (req, res) => {
     const offenderKeyworkers = offenderKeyworkerResponse.data;
     log.debug({ data: offenderKeyworkers }, 'Response from getOffenders request');
 
-    for (const row of data) {
-      const details = offenderKeyworkers.filter(d => d.offenderNo === row.offenderNo);
-      if (details.length >= 1) {
-        const detail = details[0];
-        row.staffId = detail && detail.staffId;
-      }
+    offenderResults = applyAllocationStatusFilter(allocationStatus, offenderResults, offenderKeyworkers); //adjust results if filtering by unallocated
+
+    if (offenderResults.length > 0) {
+      req.data = getOffenderNoArray(offenderResults);
+      const sentenceDetailListResponse = await elite2Api.sentenceDetailList(req, res);
+      const allReleaseDates = sentenceDetailListResponse.data;
+      log.debug({ data: allReleaseDates }, 'Response from sentenceDetailList request');
 
-      if (row.staffId) {
-        const kw = keyworkerData.find(i => i.staffId === row.staffId);
-        if (kw) { // eslint-disable-line max-depth
-          row.keyworkerDisplay = `${properCaseName(kw.lastName)}, ${properCaseName(kw.firstName)}`;
-          row.numberAllocated = kw.numberAllocated;
-        } else {
-          await common.addMissingKeyworkerDetails(req, res, row);
+      const csraListResponse = await elite2Api.csraList(req, res);
+      const allCsras = csraListResponse.data;
+      log.debug({ data: allCsras }, 'Response from csraList request');
+
+      for (const row of offenderResults) {
+        const details = offenderKeyworkers.filter(d => d.offenderNo === row.offenderNo);
+        if (details.length >= 1) {
+          const detail = details[0];
+          row.staffId = detail && detail.staffId;
         }
-      }
 
-      common.addCrsaClassification(allCsras, row);
-      common.addReleaseDate(allReleaseDates, row);
+        if (row.staffId) {
+          const kw = keyworkerData.find(i => i.staffId === row.staffId);
+          if (kw) { // eslint-disable-line max-depth
+            row.keyworkerDisplay = `${properCaseName(kw.lastName)}, ${properCaseName(kw.firstName)}`;
+            row.numberAllocated = kw.numberAllocated;
+          } else {
+            await common.addMissingKeyworkerDetails(req, res, row);
+          }
+        }
+
+        common.addCrsaClassification(allCsras, row);
+        common.addReleaseDate(allReleaseDates, row);
+      }
     }
   }
   return {
     keyworkerResponse: keyworkerData,
-    offenderResponse: data,
+    offenderResponse: offenderResults,
     partialResults: partialResults
   };
 };
 
+
+const applyAllocationStatusFilter = function (allocationStatus, currentOffenderResults, offenderKeyworkers) {
+  let offenderResults = currentOffenderResults;
+  if (allocationStatus === "unallocated") {
+    offenderResults = offenderResults.filter(offender => !offenderKeyworkers.find(keyWorker => keyWorker.offenderNo === offender.offenderNo));
```

Comment:
can you just filter on whether the offenderResults row has a staff id? maybe im confused

Reply:
i thought about that earlier, yes either would work

Reply:
I've remembered the reason I couldn't use staffId. It's not until we are iterating through the collection that we have the staffId information and it was cleaner to do the filtering before the main decorating iteration to avoid having to remove elements when iterating.